backend/README.md

```diff
@@ -53,6 +53,27 @@ can issue GraphQL requests or access GraphQL Playground in the browser.
 
 ![GraphQL Playground](../.gitbook/assets/graphql-playground.png)
 
+### Database Indices and Constraints
+
+Database indices and constraints need to be created when the database and the
+backend is running:
+
+{% tabs %}
+{% tab title="Docker" %}
+```bash
+docker-compose exec backend yarn run db:migrate init
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-01-20T11:54:17Z" title="Monday, January 20th 2020, 12:54:17 pm +01:00">Jan 20, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -53,6 +53,27 @@ can issue GraphQL requests or access GraphQL Playground in the browser.
 
 ![GraphQL Playground](../.gitbook/assets/graphql-playground.png)
 
+### Database Indices and Constraints
+
+Database indices and constraints need to be created when the database and the
+backend is running:
+
+{% tabs %}
+{% tab title="Docker" %}
+```bash
+docker-compose exec backend yarn run db:migrate init
```


I'm getting this error when running this command:
```
Error: ERROR_TRANSACTION_FAILED
    at new TransactionError (/nitro-backend/node_modules/neode/build/TransactionError.js:21:126)
    at /nitro-backend/node_modules/neode/build/index.js:456:33
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
    at Store.init (/nitro-backend/src/db/migrate/store.js:18:7) {
  errors: [
    {
      query: 'CREATE CONSTRAINT ON (model:Location) ASSERT model.id IS UNIQUE',
      params: {},
      error: [Neo4jError]
    },
    {
      query: 'CREATE CONSTRAINT ON (model:Donations) ASSERT model.id IS UNIQUE',
      params: {},
      error: [Neo4jError]
    },
    {
      query: 'CREATE CONSTRAINT ON (model:Report) ASSERT model.id IS UNIQUE',
      params: {},
      error: [Neo4jError]
    }
  ]
}
```

do I need to delete all indexes on the database before running this command? or does this error message mean something more to you?

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-01-21T10:10:27Z" title="Tuesday, January 21st 2020, 11:10:27 am +01:00">Jan 21, 2020</time>_

---


looking into this further @roschaefer ... this has nothing, it seems, to do with `Docker` vs. local setup, but production vs seed data.
It appears there is an error thrown when setting up the unique constraints on `Location` and then the rest of the 4 :point_up_2: don't get run. I ran them individually in the `neo4j` browser and got this error for `Location`, the other 4 were successful.

```
Unable to create CONSTRAINT ON ( location:Location ) ASSERT location.id IS UNIQUE:
Both Node(48981) and Node(186812) have the label `Location` and property `id` = 'place.4463223470618060'
```
Looking this `Location` up by it's id results in 4 identical `Locations` in production data, that look like
```
{
  "lng": 48.06028,
  "nameES": "Villingen-Schwenningen",
  "nameFR": "Villingen-Schwenningen",
  "nameIT": "Villingen-Schwenningen",
  "nameEN": "Villingen-Schwenningen",
  "type": "place",
  "namePT": "Villingen-Schwenningen",
  "nameDE": "Villingen-Schwenningen",
  "nameNL": "Villingen-Schwenningen",
  "name": "Villingen-Schwenningen",
  "namePL": "Villingen-Schwenningen",
  "id": "place.4463223470618060",
  "lat": 8.45861
}
```

I guess we don't really need to run `init` in production, right? should we delete these duplicate records anyways? 
Looking at our db schema in production confirms that we do no have any `CONSTRAINT` `on `Location`, so I guess we want to set that up?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-01-21T14:41:04Z" title="Tuesday, January 21st 2020, 3:41:04 pm +01:00">Jan 21, 2020</time>_

---


@mattwr18 yes, we want to set that up. I used `neode` to setup unique constraints and if that fails on production data, then we had one constraint missing on `production`. Maybe we can run a migration before we setup the constraints. Sounds odd but is the best solution that comes to my mind right now.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-01-22T12:28:18Z" title="Wednesday, January 22nd 2020, 1:28:18 pm +01:00">Jan 22, 2020</time>_

---


I've added a migration that runs first... but we would need to run `yarn db:migrate up` before running `yarn db:migrate init`... otherwise the latter will fail.
Also, I changed the timestamp of the `merge_duplicate_location_nodes` so that it run before the `merge_duplicate_user_nodes`, but actually that's not important.
We have talked about adding the migrations we have already run on production data to this folder so that we have a comprehensive list of every migration we have run... should those timestamps be altered as well to appear at the top of the list and be run first?
Also, when I added the unique constraints on `Migration` node, I realized it failed as it was creating multiple nodes for the same migration.
I've updated it to `MERGE` by title, then `ON CREATE SET` the properties and changed the timestamp to `createdAt` so that we have a record of when the migration was run in production, and it's human readable.